File: 15.cpp
class Solution {
private:
    struct Node {
        int a;
        int b;
        int c;
        Node (int a, int b, int c) {
            if (a > b) {
                swap(a, b);
            } 
            if (b > c) {
                swap(b, c);
            }
            if (a > b) {
                swap(a, b);
            }
            this->a = a;
            this->b = b;
            this->c = c;
        }
        bool friend operator < (const Node & node0, const Node & node) {
            if (node0.a == node.a) {
                if (node0.b == node.b) {
                    return (node0.c < node.c);
                } else {
                    return (node0.b < node.b);
                }
            } else {
                return (node0.a < node.a);
            }
        }
    };
public:
    int low_search(vector<int>& nums, int target) {
        int low = 0;
        int high = nums.size() - 1;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (nums[mid] >= target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
    int high_search(vector<int>& nums, int target) {
        int low = 0;
        int high = nums.size() - 1;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (nums[mid] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return high;
    }
    int search(vector<int>& nums, int target, int selected1, int selected2) {
        int low = low_search(nums, target);
        int high = high_search(nums, target);
        for (int i = low; i <= high; i++) {
            if (i != selected1 && i != selected2) {

                return i;
            }
        }
        return -1;
    }
    vector<vector<int>> threeSum(vector<int>& nums) {
        set<Node> sets;
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size(); i++) {
            for (int j = i + 1; j < nums.size(); j++) {
                int a = nums[i];
                int b = nums[j];
                int c = -(nums[i] + nums[j]);
                int pos = search(nums, c, i, j);
                if (pos != -1) {
                    sets.insert(Node(a, b, c));
                }
            }
        }
        vector<vector<int>>  res;
        for (set<Node>::iterator it = sets.begin(); it != sets.end(); it++) {
            vector<int> item;
            item.push_back((*it).a);
            item.push_back((*it).b);
            item.push_back((*it).c);
            res.push_back(item);
        }
        return res;
    }
};

File: 15.py
"""
大数据不能通过
"""
def low_search(nums, target):
    start = 0
    end = len(nums) - 1
    while start <= end:
        mid = (start + end) // 2
        if nums[mid] >= target:
            end = mid - 1
        else:
            start = mid + 1
    return start
def high_search(nums, target):
    start = 0
    end = len(nums) - 1
    while start <= end:
        mid = (start + end) // 2
        # print(mid)
        if nums[mid] > target:
            end = mid - 1
        else:
            start = mid + 1
    return end
class Solution:
    def search(self, nums, target, selected):
        low = low_search(nums, target)
        high = high_search(nums, target)
        self.cache_high[target] = high
        for i in range(low, high + 1):
            if i not in selected:
                return i
        return -1
    
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = set()
        nums = sorted(nums)
        nums_size = len(nums)
        for i in range(nums_size):
            for j in range(i+1, nums_size):
                pos = self.search(nums, -(nums[i] + nums[j]), [i, j])
                if pos != -1:
                    res.add(tuple(sorted([nums[i], nums[j], nums[pos]])))
        return sorted(list(res))